File: HW1/Codes/Kas_ilke.go
package main
import "fmt"
var globalvar bool = false
func function1()bool{
    fmt.Println("function 1 is evaluated")
    return true
}
func function2()bool{
    fmt.Println("function 2 is evaluated")
    return true
}
func function3()bool{
    fmt.Println("function 3 is evaluated")
    return true
}
func function4()bool{
    fmt.Println("function 4 is evaluated")
    return false
}
func function5()bool{
    fmt.Println("function 5 is evaluated")
    return false
}
func function6()bool{
    fmt.Println("function 6 is evaluated")
    return false
}
func function7() bool{
    globalvar = true;
    fmt.Println("function 7 is evaluated");
    return true;
}
func main() {
    fmt.Println("------------1.	Boolean operators provided---------------")
    T := true
    F := false
    fmt.Println("All combination of &&(AND) operator\n")
    fmt.Println("F&&F:", F&&F)
    fmt.Println("F&&T:", F&&T)
    fmt.Println("T&&F:", T&&F)
    fmt.Println("T&&T:", T&&T)
    fmt.Println("All combination of ||(OR) operator\n")
    fmt.Println("F&&F:", F||F)
    fmt.Println("F&&T:", F||T)
    fmt.Println("T&&F:", T||F)
    fmt.Println("T&&T:", T||T)
    fmt.Println("All combination of !(NOT) operator\n")
    fmt.Println("F:", F)
    fmt.Println("!F:", !F)
    fmt.Println("T:", T)
    fmt.Println("!T:", !T)
    fmt.Println("------------2.	Data types for operands of Boolean operators --------")
    
    if(true){
        fmt.Println("print this")
    }
    if(false){
        fmt.Println("do not print this")
    }
    a:= 13
    c:= a % 2 == 1
    fmt.Println("c:", c)
    b := a <= 9;
    fmt.Println("b:", b)
    d:= true && function1()
    fmt.Println("d:", d)
    fmt.Println("------------3.	Operator precedence rules--------")
    fmt.Println("! false && false: ",!false && false)
    fmt.Println("! true || true:", !true || true)
    fmt.Println("true || true && false:", true || true && false)
    fmt.Println("------------4. Operator associativity rules--------")
    and:=function2() && function1() && function3()
    fmt.Println("and:",and)
    or:=function5() || function6() || function4()
    fmt.Println("or:",or)
    if !!!function4(){
    fmt.Println("Right associativity")
    }
    fmt.Println("------------5.	Operand evaluation order--------")
    //code segment1
    if(globalvar && function7()){
        fmt.Println("Side effect of function 7-1")
    }
    globalvar = false;
     if(function7()&&globalvar){
        fmt.Println("Side effect of function 7-2")
    }
    //code segment2
    fmt.Println("true && true || false && false: ",true && true || false && false)
    fmt.Println("true && (true || false) && false: ",true && (true || false) && false)
    fmt.Println("------------6. Short-circuit evaluation--------")
    globalvar = false;
    if(globalvar && function7()){
        fmt.Println("Side effect of function 7-1")
    }
    fmt.Println("1- global is ", globalvar)
    globalvar = false;
     if(function7()&&globalvar){
        fmt.Println("Side effect of function 7-2")
    }
     fmt.Println("2- global is ", globalvar)
}

File: HW1/Codes/Kas_ilke.c
#include <stdio.h>
int global = 0;
int function()
{
    printf("Function is executed\n");
   return 1;
}
int function1()
{
    printf("function 1 is evaluated\n");
    return 1;
}
int function2()
{
    printf("function 2 is evaluated\n");
    return 1;
}
int function3()
{
    printf("function 3 is evaluated\n");
    return 1;
}
int function4()
{
    printf("function 4 is evaluated\n");
    return 0;
}
int function5()
{
    printf("function 5 is evaluated\n");
    return 0;
}
int function6()
{
    printf("function 6 is evaluated\n");
    return 0;
}
int function7()
{
    global++;
    printf("function 7 is evaluated\n");
    return 1;
}
int main()
{
    printf("------------------------1.Boolean operators provided---------------------------\n");
    int T= 1;
    int F= 0;
    //AND OPERATOR
    printf("All combination of &&(AND) operator\n");
    if(F&&F)
    {
        printf("F&&F is true\n");
    }
     else
    {
        printf("F&&F is false\n");
    }
      if(F&&T)
    {
        printf("F&&T is true\n");
    }
     else
    {
        printf("F&&T is false\n");
    }
    if(T&&F)
    {
        printf("T&&F is true\n");
    }
    else
    {
        printf("T&&F is false\n");
    }
    if(T&&T)
    {
        printf("T&&T is true\n");
    }
     else
    {
        printf("T&&T is false\n");
    }
  
    //OR OPERATOR
    printf("All combination of ||(OR) operator\n");
    if(F||F)
    {
        printf("F||F is true\n");
    }
     else
    {
        printf("F||F is false\n");
    }
      if(F||T)
    {
        printf("F||T is true\n");
    }
     else
    {
        printf("F||T is false\n");
    }
    if(T||F)
    {
        printf("T||F is true\n");
    }
    else
    {
        printf("T||F is false\n");
    }
    if(T||T)
    {
        printf("T||T is true\n");
    }
     else
    {
        printf("T||T is false\n");
    }
    //NOT operator
    printf("All combination of !(NOT) operator\n");
    if(T)
    {
        printf("T is true\n");
    }
    else
    {
        printf("T is false\n");
    }
      if(!T)
    {
        printf("!T is true\n");
    }
    else
    {
        printf("!T is false\n");
    }
      if(F)
    {
        printf("F is true\n");
    }
    else
    {
        printf("F is false\n");
    }
      if(!F)
    {
        printf("!F is true\n");
    }
    else
    {
        printf("!F is false\n");
    }
    printf("-----------------2.Data types for operands of Boolean operators--------------\n");
    //Code segment 1
    int var1 = 0;
    int var2 = 8; 
    float var3 = 7.89;
    int var4 = -8;
    char c = 'c';
    if(var1)
    {
        printf("%d is true\n",var1);
    }
    else
    {
        printf("%d is false\n",var1);
    }
     if(var2)
    {
        printf("%d is true\n",var2);
    }
    else
    {
        printf("%d is false\n",var2);
    }
     if(var3)
    {
        printf("%f is true\n",var3);
    }
    else
    {
        printf("%f is false\n",var3);
    }
    if(c)
    {
        printf("%c is true\n",c);
    }
    else
    {
        printf("%c is false\n",c);
    }
     if(var4)
    {
        printf("%d is true\n",var4);
    }
    else
    {
        printf("%d is false\n",var4);
    }
    if(var2+var4)
    {
        printf("%d + (%d) is true\n",var2,var4);
    }
    else
    {
        printf("%d + (%d) is false\n",var2,var4);
    }
    //code segment2
    var2 = 5;
    int var5= 10;
    var4 = 12;
   if( !(var2 + var5))
   {
       printf("!(%d + %d) is true\n",var2,var5);
   }
   else
   {
        printf("!(%d + %d) is false\n",var2,var5);
   }
   if((var2 < var5) || (var2 > var4) )
   {
       printf("Relational as operand\n");
   }
   int rel_operand = (var2 < var5);
   int rel_operand2 = (var2 > var4);
   printf("var2 < var5 is %d\n",rel_operand);
   printf("var2 > var4 is %d\n",rel_operand2);
   //code segment3
   T = 1;
   if(function() && T)
   {
       printf("Function can be used as operand\n");
   }
 printf("-------------3.Operator precedence rules----------------\n");
   //code segment 1
   if(!0 && 0)
   {
       printf("!0 && 0 is true\n");
   }
   else
   {
       printf("!0 && 0 is false\n");
   }
   //code segment 2
   if(!1 || 1)
   {
       printf("!1 || 1 is true\n");
   }
   else
   {
       printf("!1 || 1 is false\n");
   }
   //code segment3
   if(1 || 1 && 0 )
   {
       printf("1 || 1 && 0 is true\n");
   }
   else
   {
       printf("1 || 1 && 0 is false \n");
   }
   printf("-------------4.	Operator associativity rules----------------\n");
   //code segment 1
   if(function2()&&function1()&&function3())
   {
    printf("Left to right\n");
   }
   if(function5()||function6()||function4())
   {
    printf("Do not print this\n");
   }
   if(!!!function4())
   {
       printf("Right associativity\n");
   } 
   printf("-------------5.	Operand evaluation order----------------\n");
     //code segment 1
     if(global && function7())
   {
    printf("Side effect of function 7-1\n");
   }
   global = 0;
   if(function7() && global)
   {
    printf("Side effect of function 7-2\n");
   }
   //code segment 2
   if(1 && 1 || 0 && 0)
   {
       printf("1 && 1 || 0 && 0 is true \n");
   }
   else
    {
       printf("1 && 1 || 0 && 0 is false \n");
   }
   if(1 && (1 || 0) && 0)
   {
       printf("1 && (1 || 0 )&& 0 is true \n");
   }
    else
    {
       printf("1 && (1 || 0) && 0 is false \n");
   }
     printf("-------------6.	Short-circuit evaluation----------------\n");
       //code segment 1
     global = 0;
    if(global && function7())
   {
    printf("Side effect of function 7-1\n");
   }
   printf("1- global is: %d\n", global);
   global = 0;
   if(function7() && global)
   {
    printf("Side effect of function 7-2\n");
   }
   printf("2- global is: %d\n", global);
   
    return 0;
}